pyk/src/pyk/proof/proof.py

```diff
@@ -303,31 +314,169 @@ def lines(self) -> list[str]:
+    as available, and `Proof.commit()` results as they become available, and get new steps with
+    `Proof.get_steps()` and submit to thread pool.
+
+    :param P: Type of proof to be advanced in parallel
+    :param PS: Proof step: data required to perform a step of the proof.
+    :param SR: Step result: data produced by executing a PS with `Prover.step_proof` used to update the `Proof`
+
+    :param create_prover: Function which creates a new `Prover`. These provers must not reference any shared
+        data to be written during `parallel_advance_proof`, to avoid race conditions.
+    :param max_iterations: Maximum number of steps to take
+    :param fail_fast: If the proof is failing after finishing a step, halt execution even if there are
+        still available steps
+    :param max_workers: Maximum number of worker threads the pool can spawn
+    """
+    pending: set[Future[Any]] = set()
+    explored: set[PS] = set()
+    iterations = 0
+
+    main_prover = create_prover()
+
+    main_prover.init_proof(proof)
+
+    with ProverPool[P, PS, SR](create_prover=create_prover, max_workers=max_workers) as pool:
+
+        def submit_steps(_steps: Iterable[PS]) -> None:
+            for step in _steps:
+                if step in explored:
+                    continue
+                explored.add(step)
+                future: Future[Any] = pool.submit(step)  # <-- schedule steps for execution
+                pending.add(future)
+
+        submit_steps(proof.get_steps())
+
+        while True:
+            if len(pending) == 0:
+                break
+            done, _ = wait(pending, return_when='FIRST_COMPLETED')
+            future = done.pop()
+            proof_results = future.result()
+            for result in proof_results:
+                proof.commit(result)
+            proof.write_proof_data()
+            iterations += 1
+            if max_iterations is not None and max_iterations <= iterations:
+                break
+            if fail_fast and proof.failed:
+                _LOGGER.warning(f'Terminating proof early because fail_fast is set: {proof.id}')
+                break
+            submit_steps(proof.get_steps())
+            pending.remove(future)
+
+        if proof.failed:
+            proof.failure_info = main_prover.failure_info(proof)
+        proof.write_proof_data()
 
 
 class Prover(Generic[P, PS, SR]):
-    kcfg_explore: KCFGExplore
+    """Abstract class which advances `Proof`s with `init_proof()` and `step_proof()`
+    :param P: Type of proof this `Prover` operates on
+    :param PS: Proof step: data required to perform a step of the proof.
+    :param SR: Step result: data produced by executing a PS with `Prover.step_proof` used to update the `Proof`
+    """
 
-    def __init__(self, kcfg_explore: KCFGExplore):
-        self.kcfg_explore = kcfg_explore
+    @abstractmethod
+    def shutdown(self) -> None: ...
```

Comment:
Consider making this class a `ContextManager[Prover]` instead.

Reply:
I've made it a `ContextManager` as well, but I'm still having to use this `shutdown` method (I renamed it to `close`), because `ProverPool` keeps `Prover`s in a list and we do not want to close those `Prover`s until the whole `ProverPool` is closed, similar to `KoreServerPool`. Is there a way to better use the `ContextManager` feature under these circumstances?